articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
@wjwwood is "teardown" mean the node has cleaned up, but is still
publishing state? I got confused by the thread activity today and want to
make sure that the status showing the error is still being published to the
system.

On Thu, Sep 17, 2015 at 4:34 PM, William Woodall <notifications@github.com>
wrote:

> In articles/node_lifecycle.md
> <https://github.com/ros2/design/pull/34#discussion_r39795313>:
>
> > +
> > +This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
> > +
> > +Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
> > +
> > +To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
> > +
> > +If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
> > +
> > +### Fatal Error state
> > +
> > +Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
> > +
> > +Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
> > +
> > +On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
>
> @doudou <https://github.com/doudou> Let me try to summarize your
> suggestions:
>
>    - Remove the Fatal Error state in this document, because:
>       - The only way to recover from a "fatal error" is to destroy the
>       node and restart, so it's effectively the same as the destroyed state.
>       - This would mean, in my estimation, that the two onError
>       transitions would instead go directly to the destroyed state.
>    - Remove the idea of an "Exception" state since nodes should just be
>    restarted in the event of any issue.
>       - You supported this by arguing that RTT has an exception state,
>       but you do not believe you've ever seen anyone use it effectively.
>
> Is that accurate? I'm just trying to nail down the argument and get an
> actionable proposal out of it.
>
> I think I've as much as agreed with you on the first point, as I
> previously said:
>
> I wasn't aware this [the onError transition] was what was called in the
> event of a fatal error. I'd think that would be more aptly named cleanup or
> onShutdown. It's not clear to me that this should be any different than a
> destructor (or a destroy function) on the node.
>
> Basically I'm say if there is a non-recoverable error, then we should go
> directly to tear down (what ever we call that). What we do after that is
> out of scope, whether that's tear down the process and the other nodes in
> it or try to restart the node that was destroyed.
>
> For the second point, I have to disagree. I can think of a lot of
> situation where error recovery is more complicated than close and reopen. I
> can give some anecdotes:
>
>    - My ax2550 driver: The ax2550 motor driver has two modes, one is
>    configuration and one is command mode. When you open the serial device you
>    send different character to get in different modes. My node first tries to
>    connect in command mode, but if the command mode is not setup correctly
>    (setup for R/C mode rather than serial mode) then I close the connection
>    and reopen it in configuration mode, set the right configuration, and then
>    open it again in command mode. You can argue that it could not try to do
>    the reconfiguration and the user should do it separately or that it should
>    do this all without transitioning to an error state or it should always
>    open in configuration mode first to assert the right config, but those
>    would all be responses to arbitrary constraints of the system, which I
>    think is bad.
>    - My segway rmp driver: Sometimes, if not properly disconnected
>    before, the segway will still be publishing telemetry and you can start in
>    the middle of a telemetry message and that messes things up, so you can
>    close it and reopen it. However, if you close and reopen the device too
>    quickly then you can cause other issues, so I always delayed for a few
>    hundred milliseconds before trying again. Now you can argue that this is
>    the devices fault or that I can do the delay in the destructor or something
>    like that, but it probably be better to just have the node (and not the
>    supervisor) do the delay in an error state before telling the supervisor to
>    try the transition again.
>
> These are both anecdotes and specific to drivers, but just because you've
> not experienced a use case for this or you can't imagine a use case for it
> doesn't mean it's not useful. I'd be more interested in a different pattern
> which provides the same utility but in a different way. Simply stating that
> it's uncommon is not a good argument against it, in my opinion.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/design/pull/34/files#r39795313>.
>